### ISSUE TITLE
Relational of complex numbers should not raise an exception

### ISSUE THREAD

**Issue by Upabjojr** [sympy/sympy#13989]
```
In [1]: I > 0
---------------------------------------------------------------------------
TypeError: Invalid comparison of complex I
```

I believe that relational expressions of complex numbers should just remain unevaluated. Raising exceptions is a bad paradigm.

**Comment by jksuom** [comment]
I don't think that returning comparisons of manifestly complex numbers (like `I`) unevaluated would be a good idea. Their appearance almost certainly means that there is an error somewhere, and an exception is an effective way of alerting the user.

**Comment by smichr** [comment]
If for pedagogical purposes one wanted to show what can't be determined unambiguosly with SymPy, symbol hackery could be used as in "Symbol("I") < 1". Let's just close this issue as a WontFix since if there is nothing to be done with such an expression and (as @jksuom says) it likely represents an error in user intention. It is more likely to catch one's attention rather than just returning an unevaluated inequality which is very common to see when symbols are involved.

**Comment by asmeurer** [comment]
The general SymPy paradigm is to raise an exception if an assumption check gives False and remain unevaluated if it gives None or True. For example, `MatrixSymbol('x', n, n)[S(1)/2]` raises an exception but `MatrixSymbol('x', n, n)[i - S(1)/2]` does not (because `i` could be 1.5). We don't always do a check, and sometimes where you might expect a check there isn't one, like apparently you can do a summation from non-integer indices. In those cases there is a useful mathematical extension of the domain. 

For inequalities there is no meaningful extension to nonreals. So I think a type check and fail for `is_real == False` is fine here. 

The only other thing that makes sense to me is to make all nonreal inequalities return False. In this paradigm, complex numbers form a poset (partially ordered set) where reals are ordered in the usual way, and nonreals aren't comparable with anything other than themselves (so `I >= I` would be True). With this, `x > y` means "x and y are comparable and x > y". I don't particularly like that paradigm because it breaks most people's expectations of how > should work, but it is logically consistent, and there is real math behind it (posets).

**Comment by skirpichev** [comment]
> For inequalities there is no meaningful extension to nonreals.

Of course, there are, in mathematics (btw, Python doesn't assume something special about "<").  And some such extensions used in SymPy for sure, e.g. for ordering roots of polynomials.

There is no extension, which turn C into an ordered field.

> it breaks most people's expectations of how > should work

Indeed.  Probably, most users would expect totality.

**Comment by Upabjojr** [comment]
I'm a bit skeptical of raising exceptions because it might be a potential problem for scripts where you'd like the code not to stop. The potentiality of exceptions forces users to wrap everything in `try ... except` blocks.

**Comment by asmeurer** [comment]
An exception in this case probably indicates a bug (either in SymPy or in the user's code). Although to be fair, unlike the TypeError from bool(Relational), it isn't clear to me that such a bug would propagate to wrong results if not raised as an exception. Most likely it would just result in the user getting a meaningless relational in their final expression.

**Comment by asmeurer** [comment]
I've discussed this on other issues before, but it's hard to find them. Since this is the main issue for this, I will state my most recent opinion on this. 

I think relational expressions should never raise, as the OP states. An expression like `I > 0` should be False. I'm not sure if it's more prudent to keep it unevaluated or to actually evaluate it. It should evaluate to False either way inside of the assumptions system. 

There's no contradiction with `(I > 0) == False` and `(I < 0) == False`. It amounts to `x > y` meaning "`x` and `y` are comparable and x - y is positive". This would make `x > y` fully equivalent to `Q.positive(x - y)` or `(x - y).is_positive`. Things like `Implies(x > y, Not(y <= x))` work fine as long as they are written carefully. That fact is true but the converse `Implies(Not(x > y), y <= x)` is not. This works because an implication with a False antecedent is still True, so there is no contradiction with I > 0 and I < 0 both being False. 

The current behavior of raising an exception is very problematic. It means that code cannot create unevaluated boolean expressions without carefully checking if the result raises an exception. It's very easy to have situations where an expression in an inequality could raise exceptions. 

Another question is if the definition of `x > y` should be "`x` and `y` are comparable and `x - y` is positive" or "`x - y` is comparable and `x - y` is positive".  Right now there is a lot of code that carefully tries to allow `x + I > 0` (because `x` could be `-I`), but disallow `x > I`). It's perhaps more mathematically "correct" this way, but it would be a lot simpler to just make `x > y` completely equivalent to `Q.positive(x - y)`. Currently this is done because the noncomparable case raises an exception, but if it instead evaluated to False. On the other hand, the former definition is required to make `I + 1 > I` not be `True`.

**Comment by oscarbenjamin** [comment]
To be clear here when we say that `x < y` I think that non-Expr classes can be free to define the operators `<`, `>`, `<=` and `>=` in different ways so e.g. we could define something like `Interval(0, 1) < Reals == True`. I would expect that when comparing instances of `Set` with these operators the unevaluated result would be something like `ProperSubSet(x, y)` rather than `LessThan(x, y)`. For Expr classes `x < y` should be equivalent to `LessThan(x, y)`. I think that `LessThan` should require its arguments to be `Expr`. I also think that the meaning of `LessThan(x, y)` should be explicitly restricted to the extended real sense of comparison and not to any other mathematical concept of orderings that might in some contexts use the symbol `<`.

With the above clarification I will refer to `x < y` on the understanding that it means `LessThan(x, y)` and only applies to `x` and `y` as instances of `Expr`. Also when you say that `x < y` should mean "x and y are comparable" I interpret that as "x and y are extended real" so you present two possibilities:

1. `x > y` is True if x and y are extended real and x - y is (extended) positive and False otherwise
2. `x > y` is True if x - y is (extended) positive and False otherwise

I think more users/authors have good applications for definition 1. rather than 2. I personally can't think of any situation where I would write `x > y` and intend it to mean definition 2. Being able to canonicalise `x > y` to `Q.positive(x - y)` is useful on the implementation side but I don't think it's the behaviour anyone really wants.

The concern from @jksuom is valid: the exceptions raised by complex comparisons are very useful in debugging. Raising exceptions is awkward though when it happens as part of symbolic evaluation. Having inequalities evaluate to False for non-real arguments can naturally make sense out of e.g. something like this:
```julia
In [42]: summation(x**n/n, (n, 1, oo))                                                                                                                        
Out[42]: 
⎧-log(1 - x)  for x ≥ -1 ∧ x < 1
⎪                               
⎪   ∞                           
⎪  ____                         
⎪  ╲                            
⎪   ╲    n                      
⎨    ╲  x                       
⎪    ╱  ──        otherwise     
⎪   ╱   n                       
⎪  ╱                            
⎪  ‾‾‾‾                         
⎪ n = 1                         
⎩        

In [43]: summation(x**n/n, (n, 1, oo)).subs(x, I/2)                                                                                                             
---------------------------------------------------------------------------
TypeError
...
TypeError: Invalid comparison of non-real I/2
```

I'm not sure to what extent there is a backwards compatibility problem in changing this. Clearly it would not be a backward compatible change in any strict sense but I wonder if this is more often just an annoyance to users rather than something that anyone particularly depends on.

**Comment by asmeurer** [comment]
In that case, the answer would return an unevaluated Sum, so it wouldn't be wrong. The convergence condition is written in such a way that it only gives -log(1 - x) when it is true (aside: I wonder if the algorithm that generates those conditions can be made correct for complex arguments, so that it would give `abs(x) < 1`). 

Perhaps the best way is to make it unevaluated. Then the logic algorithms in the assumptions are free to translate it to Q.positive(), but we could still make `bool(I > 0)` raise an exception, the same as it currently does for symbolic inequalities like `x < y`.

**Comment by Upabjojr** [comment]
> I think that non-Expr classes can be free to define the operators `<`, `>`, `<=` and `>=` in different ways so e.g. we could define something like `Interval(0, 1) < Reals == True`

The operator overloading used in the _sets_ module is horrible.

**Comment by asmeurer** [comment]
@Upabjojr can you expand on what it is you don't like there?

**Comment by oscarbenjamin** [comment]
I think they're mostly fine apart from using `+` for union.

**Comment by asmeurer** [comment]
I'm not a huge fan of | and & because I think they're confusing (which is why I like + for union). But I don't think they should be not implemented because of that.

**Comment by Upabjojr** [comment]
> @Upabjojr can you expand on what it is you don't like there?

I would expect < to represent the _LessThan_ operator, rather than the subset operator. That's so confusing...

**Comment by oscarbenjamin** [comment]
I guess we all have different expectations for these things.

One thing I think is that it is good to use Python operators in a way that is consistent with Python the language so e.g.:
```julia
In [2]: {1} | {2}                                                                                                                              
Out[2]: {1, 2}

In [3]: {1} & {2}                                                                                                                              
Out[3]: set()

In [4]: {1} < {1, 2}                                                                                                                           
Out[4]: True

In [5]: {1} < {3, 2}                                                                                                                           
Out[5]: False

In [8]: {3, 2} - {2}                                                                                                                           
Out[8]: {3}
```
Mathematically the use of `<` than for subsets makes sense because it is an ordering and these symbols are analogous although usually used in different contexts:
https://en.wikipedia.org/wiki/Subset#%E2%8A%82_and_%E2%8A%83_symbols

Likewise logical `|` and `&` (which I read as "or" and "and") are analogous as a lattice to union and intersection so that also makes sense to me. This correspondence of the operators is important I think because it means that the same identities are satisfied and that precedence, associativity etc should work intuitively.

Using `+` for union and `-` for complement doesn't really make sense because in general we have
```
(A - B) + B != A
(A + B) - B != A
```
Using `+` for union does not correspond to the operators used in Python the language for sets, does not correspond to standard mathematical notation, does not lead to a coherent "algebra" with `-` and is also redundant when we have `|`.

**Comment by asmeurer** [comment]
+ for union corresponds to the behavior of it on lists and tuples.  

> and is also redundant when we have |.

I'm not a fan of | because like I said it is confusing. It always takes an extra step of thinking to remember which of | or & corresponds to union or intersection. 

IMO Python should also allow + for union on sets and dicts. 

> does not lead to a coherent "algebra" 

There is a coherent algebra on sets, it just works differently than the algebra on numbers. (A U B) - B = A when A and B don't intersect.
